build.gradle

```diff
@@ -213,17 +213,19 @@ if (isReleaseVersion) {
     }
 }
 
-allprojects {
-
-    // FORCE UPGRADE OF GROOVY IN DEPENDENCIES TO GROOVY 4
+// FORCE UPGRADE OF GROOVY IN DEPENDENCIES TO GROOVY 4
+// except in projects that will be run by Gradle during the build
+configure(allprojects - project(':grails-shell') - project(':grails-bootstrap') - project(':grails-gradle-model')){
```

Comment:
☝️ This does not seem to take effect:
```console
./gradlew grails-bootstrap:dI --dependency org.apache.groovy:groovy-xml --configuration compileClasspath

> Task :grails-bootstrap:dependencyInsight
org.apache.groovy:groovy-xml:4.0.22
  Variant compile:
    | Attribute Name                 | Provided | Requested         |
    |--------------------------------|----------|-------------------|
    | org.gradle.status              | release  |                   |
    | org.gradle.category            | library  | library           |
    | org.gradle.libraryelements     | jar      | classes+resources |
    | org.gradle.usage               | java-api | java-api          |
    | org.gradle.dependency.bundling |          | external          |
    | org.gradle.jvm.environment     |          | standard-jvm      |
    | org.gradle.jvm.version         |          | 17                |
   Selection reasons:
      - Selected by rule
      - By constraint

org.apache.groovy:groovy-xml:4.0.22
+--- org.apache.groovy:groovy-bom:4.0.22
|    \--- org.apache.groovy:groovy-ant:4.0.22
|         +--- compileClasspath (requested org.codehaus.groovy:groovy-ant:3.0.22)
|         \--- org.apache.groovy:groovy-bom:4.0.22 (*)
\--- org.apache.groovy:groovy-templates:4.0.22
     +--- compileClasspath (requested org.codehaus.groovy:groovy-templates:3.0.22)
     \--- org.apache.groovy:groovy-bom:4.0.22 (*)

org.codehaus.groovy:groovy-xml:3.0.22 -> org.apache.groovy:groovy-xml:4.0.22
\--- compileClasspath

(*) - Indicates repeated occurrences of a transitive dependency subtree. Gradle expands transitive dependency subtrees only once per project; repeat occurrences only display the root of the subtree, followed by this annotation.
```
However, if you add it back in the `allprojects` block with the following condition:
```groovy
allprojects {

    // FORCE UPGRADE OF GROOVY IN DEPENDENCIES TO GROOVY 4
    // except in projects that will be run by Gradle during the build
    if (!(project.name in ['grails-bootstrap', 'grails-gradle-model', 'grails-shell'])) {
        configurations.configureEach {
            resolutionStrategy.eachDependency { DependencyResolveDetails details ->
                if (details.requested.group == 'org.codehaus.groovy' && details.requested.name != 'groovy-bom') {
                    details.useTarget(group: 'org.apache.groovy', name: details.requested.name, version: groovyVersion)
                }
            }
        }
    }
}
```
```console
./gradlew grails-bootstrap:dI --dependency org.apache.groovy:groovy-xml --configuration compileClasspath
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details

> Task :grails-bootstrap:dependencyInsight
No dependencies matching given input were found in configuration ':grails-bootstrap:compileClasspath'

./gradlew grails-bootstrap:dI --dependency org.codehaus.groovy:groovy-xml --configuration compileClasspath

> Task :grails-bootstrap:dependencyInsight
org.codehaus.groovy:groovy-xml:3.0.22
  Variant compile:
    | Attribute Name                 | Provided | Requested         |
    |--------------------------------|----------|-------------------|
    | org.gradle.status              | release  |                   |
    | org.gradle.category            | library  | library           |
    | org.gradle.libraryelements     | jar      | classes+resources |
    | org.gradle.usage               | java-api | java-api          |
    | org.gradle.dependency.bundling |          | external          |
    | org.gradle.jvm.environment     |          | standard-jvm      |
    | org.gradle.jvm.version         |          | 17                |

org.codehaus.groovy:groovy-xml:3.0.22
\--- compileClasspath
```

Reply:
Gradle can be a mystery some days.  I had something similar initially, but then wanted to transition it to more of a "gradle" way `configure(allprojects - project(':grails-shell') - project(':grails-bootstrap') - project(':grails-gradle-model'))` and did not realize it stopped actually adjusting dependencies properly.  I've switched to this implementation with the check in a method, since it is reused.